### Урок 5. Рекуррентные нейронные сети/ Домашнее задание

1. Попробуйте изменить параметры нейронной сети, работающей с датасетом imdb, либо нейронной сети, работающей airline-passengers (она прилагается вместе с датасетом к уроку в виде отдельного скрипта) так, чтобы улучшить её точность. Приложите анализ

### Запускаем базовую модель:

T4 GPU (On)

In [25]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import LSTM
from keras.datasets import imdb
from keras.utils import pad_sequences

In [26]:
# используем 10000 слов, встречающихся чаще всего
max_features = 10000

In [27]:
# обрезание текстов после данного количества слов (среди top max_features наиболее используемые слова)
maxlen = 100

In [28]:
batch_size = 128 # увеличьте значение для ускорения обучения

In [29]:
def train_nn():
  print('Построение модели...')
  model = Sequential()
  model.add(Embedding(max_features, 128))
  model.add(LSTM(128, dropout=0.4, recurrent_dropout=0.4))
  model.add(Dense(32, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))
  print('Загрузка данных...')
 #стоит попробовать использовать другие оптимайзер и другие конфигурации оптимайзеров
  model.compile(loss='binary_crossentropy',
                optimizer='Adam', # при использовании этого оптимайзера модель показывает наилучшие результаты.
                metrics=['accuracy'])
  (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
  print(len(x_train), 'тренировочные последовательности')
  print(len(x_test), 'тестовые последовательности')

  print('Pad последовательности (примеров в x единицу времени)')
  x_train = pad_sequences(x_train, maxlen=maxlen)
  x_test = pad_sequences(x_test, maxlen=maxlen)
  print('x_train shape:', x_train.shape)
  print('x_test shape:', x_test.shape)

  print('Процесс обучения...')
  history = model.fit(x_train, y_train,
            batch_size=batch_size,
            epochs=10, # увеличьте при необходимости
            validation_data=(x_test, y_test))
  score, acc = model.evaluate(x_test, y_test,
                              batch_size=batch_size)
  print("Результат при тестировании: ", float(round((score*100),2)), "%")
  print("Тестовая точность: ", float(round((acc*100),2)), "%")

In [30]:
train_nn()

Построение модели...
Загрузка данных...
17464789/17464789 [==============================] - 0s 0us/step
25000 тренировочные последовательности
25000 тестовые последовательности
Pad последовательности (примеров в x единицу времени)
x_train shape: (25000, 100)
x_test shape: (25000, 100)
Процесс обучения...
Epoch 1/10
196/196 [==============================] - 213s 1s/step - loss: 0.4419 - accuracy: 0.7832 - val_loss: 0.3702 - val_accuracy: 0.8404
Epoch 2/10
196/196 [==============================] - 217s 1s/step - loss: 0.2810 - accuracy: 0.8865 - val_loss: 0.3618 - val_accuracy: 0.8395
Epoch 3/10
196/196 [==============================] - 217s 1s/step - loss: 0.2257 - accuracy: 0.9098 - val_loss: 0.4482 - val_accuracy: 0.8450
Epoch 4/10
196/196 [==============================] - 217s 1s/step - loss: 0.1799 - accuracy: 0.9290 - val_loss: 0.3948 - val_accuracy: 0.8428
Epoch 5/10
196/196 [==============================] - 213s 1s/step - loss: 0.1420 - accuracy: 0.9471 - val_loss: 0.5156 -

##
### Upgrade to model2:

In [31]:
# используем 10000 слов, встречающихся чаще всего
max_features = 10000

In [32]:
# обрезание текстов после данного количества слов (среди top max_features наиболее используемые слова)
maxlen2 = 200  # was 100

In [33]:
batch_size2 = 256 # was 128

In [34]:
def train_nn2():
  print('Построение модели_2...')
  model2 = Sequential()
  model2.add(Embedding(max_features, 256))  # was changed
  model2.add(LSTM(128, dropout=0.4, recurrent_dropout=0.4))
  model2.add(Dense(32, activation='relu'))
  model2.add(Dense(1, activation='sigmoid'))
  print('Загрузка данных...')
 #стоит попробовать использовать другие оптимайзер и другие конфигурации оптимайзеров
  model2.compile(loss='binary_crossentropy',
                optimizer='SGD', # was Adam
                metrics=['accuracy'])
  (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
  print(len(x_train), 'тренировочные последовательности')
  print(len(x_test), 'тестовые последовательности')

  print('Pad последовательности (примеров в x единицу времени)')
  x_train = pad_sequences(x_train, maxlen=maxlen)
  x_test = pad_sequences(x_test, maxlen=maxlen)
  print('x_train shape:', x_train.shape)
  print('x_test shape:', x_test.shape)

  print('Процесс обучения...')
  history2 = model2.fit(x_train, y_train,
            batch_size=batch_size2,
            epochs=10, # увеличьте при необходимости
            validation_data=(x_test, y_test))
  score2, acc2 = model2.evaluate(x_test, y_test,
                              batch_size=batch_size)
  print("Результат при тестировании: ", float(round((score2*100),2)), "%")
  print("Тестовая точность: ", float(round((acc2*100),2)), "%")

In [35]:
train_nn2()

Построение модели_2...
Загрузка данных...
25000 тренировочные последовательности
25000 тестовые последовательности
Pad последовательности (примеров в x единицу времени)
x_train shape: (25000, 100)
x_test shape: (25000, 100)
Процесс обучения...
Epoch 1/10
98/98 [==============================] - 190s 2s/step - loss: 0.6932 - accuracy: 0.4926 - val_loss: 0.6932 - val_accuracy: 0.4968
Epoch 2/10
98/98 [==============================] - 199s 2s/step - loss: 0.6932 - accuracy: 0.4991 - val_loss: 0.6931 - val_accuracy: 0.5046
Epoch 3/10
98/98 [==============================] - 198s 2s/step - loss: 0.6931 - accuracy: 0.5027 - val_loss: 0.6930 - val_accuracy: 0.5125
Epoch 4/10
98/98 [==============================] - 196s 2s/step - loss: 0.6929 - accuracy: 0.5128 - val_loss: 0.6929 - val_accuracy: 0.5156
Epoch 5/10
98/98 [==============================] - 196s 2s/step - loss: 0.6929 - accuracy: 0.5155 - val_loss: 0.6929 - val_accuracy: 0.5192
Epoch 6/10
98/98 [==============================] -

**Вывод:** *неудачная конфигурация*

## Upgrade to model3:
вернём Adam

In [36]:
# обрезание текстов после данного количества слов (среди top max_features наиболее используемые слова)
maxlen3 = 200  # no changes from model2

In [37]:
batch_size3 = 256 # no changes from model2

In [38]:
def train_nn3():
  print('Построение модели_3...')
  model3 = Sequential()
  model3.add(Embedding(max_features, 256))  # was changed
  model3.add(LSTM(128, dropout=0.4, recurrent_dropout=0.4))
  model3.add(Dense(32, activation='relu'))
  model3.add(Dense(1, activation='sigmoid'))
  print('Загрузка данных...')
 #стоит попробовать использовать другие оптимайзер и другие конфигурации оптимайзеров
  model3.compile(loss='binary_crossentropy',
                optimizer='Adam', # was SGD
                metrics=['accuracy'])
  (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
  print(len(x_train), 'тренировочные последовательности')
  print(len(x_test), 'тестовые последовательности')

  print('Pad последовательности (примеров в x единицу времени)')
  x_train = pad_sequences(x_train, maxlen=maxlen)
  x_test = pad_sequences(x_test, maxlen=maxlen)
  print('x_train shape:', x_train.shape)
  print('x_test shape:', x_test.shape)

  print('Процесс обучения...')
  history3 = model3.fit(x_train, y_train,
            batch_size=batch_size3,
            epochs=10, # увеличьте при необходимости
            validation_data=(x_test, y_test))
  score3, acc3 = model3.evaluate(x_test, y_test,
                              batch_size=batch_size)
  print("Результат при тестировании: ", float(round((score3*100),2)), "%")
  print("Тестовая точность: ", float(round((acc3*100),2)), "%")

In [39]:
train_nn3()

Построение модели_3...
Загрузка данных...
25000 тренировочные последовательности
25000 тестовые последовательности
Pad последовательности (примеров в x единицу времени)
x_train shape: (25000, 100)
x_test shape: (25000, 100)
Процесс обучения...
Epoch 1/10
98/98 [==============================] - 202s 2s/step - loss: 0.4776 - accuracy: 0.7553 - val_loss: 0.3416 - val_accuracy: 0.8487
Epoch 2/10
98/98 [==============================] - 187s 2s/step - loss: 0.2790 - accuracy: 0.8853 - val_loss: 0.3618 - val_accuracy: 0.8496
Epoch 3/10
98/98 [==============================] - 199s 2s/step - loss: 0.2190 - accuracy: 0.9140 - val_loss: 0.4198 - val_accuracy: 0.8404
Epoch 4/10
98/98 [==============================] - 199s 2s/step - loss: 0.1722 - accuracy: 0.9333 - val_loss: 0.4095 - val_accuracy: 0.8444
Epoch 5/10
98/98 [==============================] - 202s 2s/step - loss: 0.1317 - accuracy: 0.9504 - val_loss: 0.5540 - val_accuracy: 0.8380
Epoch 6/10
98/98 [==============================] -

**Вывод:** *пока это лучший вариант*

## Upgrade to model4:

изменим количество и качество слоёв

In [40]:
from keras.layers import Dropout

In [41]:
def train_nn4():
  print('Построение модели_4...')
  model4 = Sequential()
  model4.add(Embedding(max_features, 256))
  model4.add(LSTM(128, dropout=0.4, recurrent_dropout=0.4))
  model4.add(Dense(64, activation='relu'))  # was 32
  model4.add(Dropout(0.4))
  model4.add(Dense(32, activation='LeakyReLU'))            # New line
  model4.add(Dropout(0.4))
  model4.add(Dense(16, activation='elu'))
  model4.add(Dropout(0.2))                                  # New line
  model4.add(Dense(1, activation='sigmoid'))

  print('Загрузка данных...')
 #стоит попробовать использовать другие оптимайзер и другие конфигурации оптимайзеров
  model4.compile(loss='binary_crossentropy',
                optimizer='Adam',
                metrics=['accuracy'])
  (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
  print(len(x_train), 'тренировочные последовательности')
  print(len(x_test), 'тестовые последовательности')

  print('Pad последовательности (примеров в x единицу времени)')
  x_train = pad_sequences(x_train, maxlen=maxlen)
  x_test = pad_sequences(x_test, maxlen=maxlen)
  print('x_train shape:', x_train.shape)
  print('x_test shape:', x_test.shape)

  print('Процесс обучения...')
  history4 = model4.fit(x_train, y_train,
            batch_size=batch_size3,
            epochs=10, # увеличьте при необходимости
            validation_data=(x_test, y_test))
  score4, acc4 = model4.evaluate(x_test, y_test,
                              batch_size=batch_size)
  print("Результат при тестировании: ", float(round((score4*100),2)), "%")
  print("Тестовая точность: ", float(round((acc4*100),2)), "%")

In [42]:
train_nn4()

Построение модели_4...
Загрузка данных...
25000 тренировочные последовательности
25000 тестовые последовательности
Pad последовательности (примеров в x единицу времени)
x_train shape: (25000, 100)
x_test shape: (25000, 100)
Процесс обучения...
Epoch 1/10
98/98 [==============================] - 192s 2s/step - loss: 0.5300 - accuracy: 0.7180 - val_loss: 0.3450 - val_accuracy: 0.8471
Epoch 2/10
98/98 [==============================] - 200s 2s/step - loss: 0.3118 - accuracy: 0.8782 - val_loss: 0.3574 - val_accuracy: 0.8509
Epoch 3/10
98/98 [==============================] - 202s 2s/step - loss: 0.2434 - accuracy: 0.9082 - val_loss: 0.4709 - val_accuracy: 0.8386
Epoch 4/10
98/98 [==============================] - 202s 2s/step - loss: 0.1905 - accuracy: 0.9281 - val_loss: 0.4344 - val_accuracy: 0.8410
Epoch 5/10
98/98 [==============================] - 200s 2s/step - loss: 0.1443 - accuracy: 0.9470 - val_loss: 0.4675 - val_accuracy: 0.8380
Epoch 6/10
98/98 [==============================] -

**Вывод:** *удалось немного повысить точность! Но **val_loss** высоковат*

### Upgrade model4 to model5
Меньше эпох и optimizer='Nadam'

In [43]:
def train_nn5():
  print('Построение модели_5...')
  model5 = Sequential()
  model5.add(Embedding(max_features, 256))
  model5.add(LSTM(128, dropout=0.4, recurrent_dropout=0.4))
  model5.add(Dense(64, activation='relu'))
  model5.add(Dropout(0.4))
  model5.add(Dense(32, activation='LeakyReLU'))
  model5.add(Dropout(0.4))
  model5.add(Dense(16, activation='elu'))
  model5.add(Dropout(0.2))
  model5.add(Dense(1, activation='sigmoid'))

  print('Загрузка данных...')
 #стоит попробовать использовать другие оптимайзер и другие конфигурации оптимайзеров
  model5.compile(loss='binary_crossentropy',
                optimizer='Nadam',
                metrics=['accuracy'])
  (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
  print(len(x_train), 'тренировочные последовательности')
  print(len(x_test), 'тестовые последовательности')

  print('Pad последовательности (примеров в x единицу времени)')
  x_train = pad_sequences(x_train, maxlen=maxlen)
  x_test = pad_sequences(x_test, maxlen=maxlen)
  print('x_train shape:', x_train.shape)
  print('x_test shape:', x_test.shape)

  print('Процесс обучения...')
  history5 = model5.fit(x_train, y_train,
            batch_size=batch_size3,
            epochs=8, # was 10
            validation_data=(x_test, y_test))
  score5, acc5 = model5.evaluate(x_test, y_test,
                              batch_size=batch_size)
  print("Результат при тестировании: ", float(round((score5*100),2)), "%")
  print("Тестовая точность: ", float(round((acc5*100),2)), "%")

In [44]:
train_nn5()

Построение модели_5...
Загрузка данных...
25000 тренировочные последовательности
25000 тестовые последовательности
Pad последовательности (примеров в x единицу времени)
x_train shape: (25000, 100)
x_test shape: (25000, 100)
Процесс обучения...
Epoch 1/8
98/98 [==============================] - 209s 2s/step - loss: 0.5013 - accuracy: 0.7415 - val_loss: 0.3457 - val_accuracy: 0.8491
Epoch 2/8
98/98 [==============================] - 202s 2s/step - loss: 0.2924 - accuracy: 0.8862 - val_loss: 0.3559 - val_accuracy: 0.8492
Epoch 3/8
98/98 [==============================] - 200s 2s/step - loss: 0.2240 - accuracy: 0.9150 - val_loss: 0.3556 - val_accuracy: 0.8466
Epoch 4/8
98/98 [==============================] - 202s 2s/step - loss: 0.1761 - accuracy: 0.9349 - val_loss: 0.4319 - val_accuracy: 0.8411
Epoch 5/8
98/98 [==============================] - 205s 2s/step - loss: 0.1353 - accuracy: 0.9515 - val_loss: 0.4984 - val_accuracy: 0.8329
Epoch 6/8
98/98 [==============================] - 204s 

**Вывод:** *удалось опять же немного повысить точность и снизить **val_loss** по сравнению с model4*

##

Model3 выглядит более-менее прилично, по сравнению с остальными.
Попробуем её улучшить...

### Upgrade model3 to model6:

In [46]:
from keras.layers import SimpleRNN

In [64]:
def train_nn6():
  print('Построение модели_6...')
  model6 = Sequential()
  model6.add(Embedding(max_features, 256))  # was changed
  model6.add(LSTM(128, dropout=0.6, recurrent_dropout=0.4))
  #model6.add(SimpleRNN(64, dropout=0.4, recurrent_dropout=0.2)) - не особо помогло
  model6.add(Dense(32, activation='relu'))
  model6.add(Dense(1, activation='sigmoid'))
  print('Загрузка данных...')
 #стоит попробовать использовать другие оптимайзер и другие конфигурации оптимайзеров
  model6.compile(loss='binary_crossentropy',
                optimizer='Nadam', # was Adam.., # Nadam - the best
                metrics=['accuracy'])
  (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
  print(len(x_train), 'тренировочные последовательности')
  print(len(x_test), 'тестовые последовательности')

  print('Pad последовательности (примеров в x единицу времени)')
  x_train = pad_sequences(x_train, maxlen=maxlen)
  x_test = pad_sequences(x_test, maxlen=maxlen)
  print('x_train shape:', x_train.shape)
  print('x_test shape:', x_test.shape)

  print('Процесс обучения...')
  history6 = model6.fit(x_train, y_train,
            batch_size=batch_size3,
            epochs=3, # увеличьте при необходимости - bad idea!
            validation_data=(x_test, y_test))
  score6, acc6 = model6.evaluate(x_test, y_test,
                              batch_size=batch_size)
  print("Результат при тестировании: ", float(round((score6*100),2)), "%")
  print("Тестовая точность: ", float(round((acc6*100),2)), "%")

In [65]:
train_nn6()

Построение модели_6...
Загрузка данных...
25000 тренировочные последовательности
25000 тестовые последовательности
Pad последовательности (примеров в x единицу времени)
x_train shape: (25000, 100)
x_test shape: (25000, 100)
Процесс обучения...
Epoch 1/3
98/98 [==============================] - 214s 2s/step - loss: 0.4939 - accuracy: 0.7545 - val_loss: 0.3693 - val_accuracy: 0.8350
Epoch 2/3
98/98 [==============================] - 207s 2s/step - loss: 0.2839 - accuracy: 0.8833 - val_loss: 0.3542 - val_accuracy: 0.8463
Epoch 3/3
196/196 [==============================] - 41s 209ms/step - loss: 0.3423 - accuracy: 0.8495
Результат при тестировании:  34.23 %
Тестовая точность:  84.95 %


**В итоге** *удалось добиться лучшей совокупности показателей модели_6 (Результат при тестировании: 34.23 %, Тестовая точность: 84.95 %) при модификации изначального набора слоёв (без добвления дополнительных), применении оптимизатора 'Nadam' и уменьшения количества эпох.* 🤕